fix(nuxt): scroll to top on dynamic routes with different params

Atinux opened the pull request:
<!---
☝️ PR title should follow conventional commits (https://conventionalcommits.org)

Please carefully read the contribution docs before creating a pull request
 👉 https://v3.nuxtjs.org/community/contribution
-->

### 🔗 Linked issue

<!-- Please ensure there is an open issue and mention its number as #123 -->

### ❓ Type of change

<!-- What types of changes does your code introduce? Put an `x` in all the boxes that apply. -->

- [ ] 📖 Documentation (updates to the documentation or readme)
- [x] 🐞 Bug fix (a non-breaking change that fixes an issue)
- [ ] 👌 Enhancement (improving an existing functionality like performance)
- [ ] ✨ New feature (a non-breaking change that adds functionality)
- [ ] ⚠️ Breaking change (fix or feature that would cause existing functionality to change)

### 📚 Description

<!-- Describe your changes in detail -->
<!-- Why is this change required? What problem does it solve? -->
<!-- If it resolves an open issue, please link to the issue here. For example "Resolves #1337" -->
Improvement over #3851 so we scroll to top when it's the same page but with different params.

Example: Going from `/movies/14` to `/movies/15` should scroll to top.

I am only using params to `/movies` to `/movies?q=test` does not scroll to top.

### 📝 Checklist

<!-- Put an `x` in all the boxes that apply. -->
<!-- If your change requires a documentation PR, please link it appropriately -->
<!-- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->

- [ ] I have linked an issue or discussion.
- [ ] I have updated the documentation accordingly.

Atinux commented on packages/nuxt/src/pages/runtime/router.options.ts:
```diff
@@ -16,11 +16,8 @@ export default <RouterConfig> {
-      (from && to && from.matched[0] !== to.matched[0]) &&
-      to.meta.scrollToTop !== false
-    ) {
+    const isDifferentRoute = from && to && (from.matched[0] !== to.matched[0] || (from.matched[0] === to.matched[0] && JSON.stringify(from.params) !== JSON.stringify(to.params)))
```
Yes but then we need to add `ohash` to the client bundle, adding some more JS 🤔

pi0 replied:
Also small perf: We need to evaluate it only when !position is met. Pushing refactor

chrisnoden commented:
I've updated to rc.13 and am not seeing that this issue is fixed. Navigating to a fresh "page" and the scroll position does not return to the top of the page.

Atinux commented:
It should, can you confirm you don't have a `app/router.options.ts` in your project?

chrisnoden commented:
I do not have that file. I removed my .nuxt and node_modules before updating to rc.13 and have also updated the @nuxtjs libs. My project works fine in all other respects (so far) but navigating (via NuxtLink) to a fresh page from, eg the footer nav, and the scroll position remains so that the footer is still in view and the actual change to the body can't be seen because it is above the fold.

I'll check it on my minimal nuxt project.

chrisnoden commented:
Confirmed. I tested in a bare bones (only nuxt rc.13) project and if my NuxtLink is at the bottom of a long page, clicking it does not cause the browser to scroll to the top. So, this bug is still not resolved.

I am unable to get rc6 to run now, but from my perspective this issue started with rc7.

misaon commented:
Same problem. Nested pages scroll to top is not implemented too.

chrisnoden commented:
Here's my minimal project which demonstrates the issue:
https://github.com/chrisnoden/nuxt3-bug

Atinux commented:
Alright I found the bug, it comes after we removed the transitions by default (PR #8463) cc @danielroe 

The hooks to wait is `page:transition:finish` by default, seems that he does not detect we removed the transitions by default.

To fix that, you can define the `app/router.config.ts`, see my fix on https://github.com/chrisnoden/nuxt3-bug/pull/1/files

Atinux commented:
Opened a PR to fix this: https://github.com/nuxt/framework/pull/8700

Really sorry about the regression 😢

chrisnoden commented:
Thanks for the prompt attention. I can confirm this fixes https://github.com/nuxt/framework/pull/8327 (also https://github.com/nuxt/framework/issues/8696)

darklow commented:
Is it possible to disable this scroll to top behaviour for `useFetch`? It messes up infinite scroll, where instead of replacement new items are added to the bottom of the page. Thanks

Neuvier commented:
I use a nested route with a parent and a children. I don't want to scroll to top when i click on a nuxtlink to display the children. It s working fine when i use a static route (ex /parent/children.vue), but when i use a dynamic route (ex /parent/[children].vue) the behavior change with a scroll to top on the children page ? the scroll to top should be the same for static and dynamic nested routes ? Is it a bug, or  i need to configure something specific ?
